### PR TITLE
DynamicParams: refactor Parameter and add support for Enums

### LINKED ISSUE #1549: DynamicParameters: Let const parameters register callbacks
### Description of the task

From slack

> Being able to register a callback with a "non-mutable" Parameter. For example in nearly all cases we don't want code to be able to set Parameter values itself, but it's totally reasonable for that code to want to listen for changes to those values and bind callbacks. Currently we can't do this because the registerCallback function is non-const, meaning the non-mutable DynamicParameters can't use that function (since DynamicParameters are const ). This may necessitate two different parameter types to enforce mutability / non-mutability rather than relying solely on const

### Acceptance criteria

- [ ] Let "const" parameters register callbacks

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description

- refactored parameter.h to remove unused clutter
- added `ContinousParameter` class to indicate numeric parameters with ranges between min/max
- added `DiscreteParameter` class to support parameters with discrete values
- disallow invalid parameter construction
- reject invalid values NOTE: we reject invalid values instead of throwing an exception to allow the the user of the params to handle that case

update:
- also allow const parameters to register callbacks

### Testing Done
unit tests for both new param classes
tested enum works with `DiscreteParameter`
tested old parameters still work (visualizer start/stop)

### Resolved Issues
Working towards generating the Example C++ Config, #1447 
Resolves #1549

### Length Justification

N/A
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/parameter/continous_parameter.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+     *
+     * @return the max of this parameter
+     */
+    const T getMax() const
+    {
+        return max_;
+    }
+
+    /**
+     * Given the value, checks if the value between the min and max value, and if so sets
```
Comment (jonathanlew) [inline src/software/parameter/continous_parameter.h]:
```suggestion
     * Given the value, checks if the value is between the min and max value, and if so sets
```

File: src/software/parameter/parameter_test.cpp [hunk 2]
```diff
@@ -1,14 +1,150 @@
+
```
Comment (jonathanlew) [inline src/software/parameter/parameter_test.cpp]: nit: remove new line

**Review by jonathanlew — COMMENTED** [review]
Looks good, just got a couple nits

File: src/software/parameter/continous_parameter.h (hunk 1 above)
Comment (akhilveeraghanta) [reply src/software/parameter/continous_parameter.h]: fixed

File: src/software/parameter/parameter_test.cpp (hunk 2 above)
Comment (akhilveeraghanta) [reply src/software/parameter/parameter_test.cpp]: fixed

**Review by jonathanlew — DISMISSED** [review]
:fire:

**Review by jonathanlew — DISMISSED** [review]

File: src/software/parameter/parameter.h [hunk 3]
```diff
@@ -103,32 +68,24 @@ class Parameter
...
         callback_functions.emplace_back(callback);
     }
 
-   private:
-    // this mutex is marked as "mutable" so that it can be acquired in a const function
+   protected:
+    // mutexes are marked as mutable so that they can be acquired in a const function
     mutable std::mutex value_mutex_;
-    std::mutex callback_mutex_;
+    mutable std::mutex callback_mutex_;
```
Comment (MathewMacDougall) [inline src/software/parameter/parameter.h]: Seems like the mutexes aren't used by the derived classes so can be made `private`

File: src/software/parameter/parameter.h [hunk 4]
```diff
@@ -103,32 +68,24 @@ class Parameter
...
-    // Store the options of this parameter
-    std::vector<T> options_;
-
     // Store the name of the parameter
     std::string name_;
 
     // A list of functions to call when a new parameter value is set
-    std::vector<std::function<void(T)>> callback_functions;
+    // marked as mutable to allow const parameters to also register callbacks
+    mutable std::vector<std::function<void(T)>> callback_functions;
```
Comment (MathewMacDougall) [inline src/software/parameter/parameter.h:86]: very cool, TIL

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looking super neato burrito

File: src/software/parameter/parameter.h [hunk 5]
```diff
@@ -90,6 +54,7 @@ class Parameter
     {
         std::scoped_lock value_lock(this->value_mutex_);
         this->value_ = new_value;
+
```
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/parameter/parameter.h:57]: was thinking a `Parameter` could still be used for "free form" values, like a free string (with no limitations on allowed values)

if we can't think of a case where that would be useful, I can easily make this `protected`

File: src/software/parameter/continous_parameter.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "software/logger/logger.h"
+#include "software/parameter/parameter.h"
+
+template <class T>
+class ContinousParameter : public Parameter<T>
+{
+    /**
+     * ContinousParameters can only be templated with numeric types
+     * we check using the is_arithmetic trait to disallow non-arithmetic types
```
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/parameter/continous_parameter.h]: It was the best naming I could come up with, open to suggestions

Was discussing with @jonathanlew offline and we came up with `BoundedParameters` and `FixedValueParameters`. 

The issue with `Numerical` and `NonNumerical` is that the current `Discrete` parameter can still have numerical values. The `Continous` parameter is strictly numerical

File: src/software/parameter/parameter.h (hunk 3 above)
Comment (akhilveeraghanta) [reply src/software/parameter/parameter.h]: good catch, done

File: src/software/parameter/parameter.h (hunk 4 above)
Comment (akhilveeraghanta) [reply src/software/parameter/parameter.h:86]: yeah its pretty sick, you have to mark` mutexes` as `mutable` if u want to use them in a const function

File: src/software/parameter/parameter.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/parameter/parameter.h:57]: Currently that's useful for things like the network config, so lets leave it.

File: src/software/parameter/continous_parameter.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/parameter/continous_parameter.h]: Maybe `Numerical` and `FixedValue`?

@garethellis0 @EvanMorcom @kelvinkoon @LiCody summoning more opinions

**Review by jonathanlew — COMMENTED** [review]
Nothing to add beyond what Akhil summarized re: bounded vs fixedvalueparams

**Review by jonathanlew — COMMENTED** [review]
.

File: src/software/parameter/continous_parameter.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/parameter/continous_parameter.h]: How about `EnumeratedParameter` and `RangedParameter`?

**Review by garethellis0 — COMMENTED** [review]
Very cool, looks like we just need to resolve the naming issue.

File: src/software/parameter/continous_parameter.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/parameter/continous_parameter.h]: Yeah that one is clear, `Ranged` and `Enumerated` make sense

[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/parameter/continous_parameter.h]: Changed to `Numeric` and `Enumerated` after offline discussion

File: src/software/parameter/parameter_test.cpp [hunk 7]
```diff
@@ -4,11 +4,160 @@
...
+    EXPECT_TRUE(test_optional_int.value());
+    EXPECT_TRUE(test_optional_int.value() == 100);
+
+    test_optional_int.setValue(std::nullopt);
+
+    EXPECT_FALSE(test_optional_int.value());
+}
+
+TEST(ParameterTest, test_continous_parameter_int)
```
Comment (jonathanlew) [inline src/software/parameter/parameter_test.cpp]:
```suggestion
TEST(ParameterTest, test_numeric_parameter_int)
```

File: src/software/parameter/parameter_test.cpp [hunk 8]
```diff
@@ -4,11 +4,160 @@
...
+{
+    NumericParameter<int> test_continous_parameter =
```
Comment (jonathanlew) [inline src/software/parameter/parameter_test.cpp]:
```suggestion
    NumericParameter<int> test_numeric_parameter =
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Some lingering name changes in tests

File: src/software/parameter/parameter_test.cpp (hunk 7 above)
Comment (akhilveeraghanta) [reply src/software/parameter/parameter_test.cpp]: good catch, fixed

File: src/software/parameter/parameter_test.cpp (hunk 8 above)
Comment (akhilveeraghanta) [reply src/software/parameter/parameter_test.cpp]: fixed and all the other cases as well

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by LiCody — APPROVED** [review]
LGTM 🎉

**Review by MathewMacDougall — APPROVED** [review]
:fire: :100:

**Pull request merged**